cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -818,13 +847,80 @@ private Location copyProgramJar() throws IOException {
     return programJarCopy;
   }
 
+  /**
+   * Creates a JAR file that contains all the plugin jars that are needed by the job. The plugin directory
+   * structure is maintained inside the jar so that MR framework can correctly expand and recreate the required
+   * structure.
+   *
+   * @return the {@link Location} for the archive file or {@code null} if there is no plugin files need to be localized
+   */
+  @Nullable
+  private Location createPluginArchive(@Nullable AdapterSpecification adapterSpec,
+                                       Id.Program programId) throws IOException {
+    if (adapterSpec == null) {
+      return null;
+    }
+
+    Set<PluginInfo> pluginInfos = adapterSpec.getPluginInfos();
+    if (pluginInfos.isEmpty()) {
```

Comment:
It's not required for adapter to have plugin. Someone can write an app template that doesn't use plugin at all.